uniffi_macros/src/export/scaffolding.rs

```diff
@@ -153,11 +153,60 @@ fn gen_ffi_function(
     // FIXME(jplatte): Use an extra trait implemented for `T: FfiConverter` as
     // well as `()` so no different codegen is needed?
     let (output, return_expr);
+    let mut extra_functions = Vec::new();
+
     match &sig.output {
         ReturnType::Default => {
             output = None;
             return_expr = rust_fn_call;
         }
+
+        ReturnType::Type(_, ty) if sig.asyncness.is_some() => {
+            // The function returns `Option<Box<…>>` so that the `FfiDefault`
+            // implementation can return `None` rather than `Box`, which could
+            // leak data in case of errors.
+            output = Some(quote! {
+               -> Option<Box<::uniffi::RustFuture<#ty>>>
+            });
+            return_expr = quote! {
+                Some(Box::new(::uniffi::RustFuture::new(
+                    async move {
+                        #rust_fn_call.await
+                    }
+                )))
+            };
+
+            let ffi_poll_ident = format_ident!("{}_poll", ffi_ident);
+            let ffi_drop_ident = format_ident!("{}_drop", ffi_ident);
+
+            // Monomorphised poll function.
+            extra_functions.push(quote! {
+                #[doc(hidden)]
+                #[no_mangle]
+                pub extern "C" fn #ffi_poll_ident(
+                    future: core::option::Option<&mut ::uniffi::RustFuture<#ty>>,
+                    waker: core::option::Option<core::ptr::NonNull<::uniffi::RustFutureForeignWakerFunction>>,
+                    waker_environment: *const ::uniffi::RustFutureForeignWakerEnvironment,
+                    polled_result: &mut <#ty as ::uniffi::FfiConverter>::FfiType,
+                    call_status: &mut ::uniffi::RustCallStatus,
+                ) -> bool {
+                    ::uniffi::ffi::uniffi_rustfuture_poll(future, waker, waker_environment, polled_result, call_status)
+                }
+            });
+
+            // Monomorphised drop function.
+            extra_functions.push(quote! {
+                #[doc(hidden)]
+                #[no_mangle]
+                pub extern "C" fn #ffi_drop_ident(
+                    future: core::option::Option<Box<::uniffi::RustFuture<#ty>>>,
```

Comment:
I don't think supporting no-std in UniFFI is very realistic in the next years so you could write `std` instead of `core`, but either way, makes sense to consistently use absolute paths (here and elsewhere):

```suggestion
                    future: ::core::option::Option<::alloc::boxed::Box<::uniffi::RustFuture<#ty>>>,
```